vms/proposervm/block.go

```diff
@@ -79,6 +80,34 @@ func (p *postForkCommonComponents) Height() uint64 {
 	return p.innerBlk.Height()
 }
 
+// Calculates a block's P-Chain epoch height based on its ancestor's epoch membership
+func nextPChainEpoch(parentPChainHeight uint64, parentEpoch block.PChainEpoch, parentTimestamp time.Time, epochDuration time.Duration) block.PChainEpoch {
+	if parentEpoch.Number == 0 {
+		// If the parent was not assigned an epoch, then the child is the first block of
+		// the initial epoch.
+		return block.PChainEpoch{
+			Height:    parentPChainHeight,
+			Number:    1,
+			StartTime: parentTimestamp,
+		}
+	}
+
+	if parentTimestamp.After(parentEpoch.StartTime.Add(epochDuration)) {
+		// If the parent crossed the epoch boundary, then it sealed the previous epoch. The child
+		// is the first block of the new epoch, so should use the parent's P-Chain height, increment
+		// the epoch number, and set the epoch start time to the parent's timestamp.
+		return block.PChainEpoch{
+			Height:    parentPChainHeight,
+			Number:    parentEpoch.Number + 1,
+			StartTime: parentTimestamp,
+		}
+	}
+	// Otherwise, the parent did not seal the previous epoch, so the child should use the parent's
+	// epoch information. This is true even if the child crosses the epoch boundary, since sealing
+	// blocks are considered to be part of the epoch they seal.
+	return parentEpoch
```

Comment:
I think we should have a unit test just to test this function. Not trying to nitpick but this line isn't covered by tests....  Of course we don't need to have every line to be covered by tests but I think this is a critical function so we probably should have a test just for it that covers all its cases.